apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/PODPipeline.ts

```diff
@@ -0,0 +1,439 @@
+   */
+  private async storeCachedPCD(
+    id: string,
+    serializedPCD: SerializedPCD
+  ): Promise<void> {
+    const key = await getHash(`${id}${this.eddsaPrivateKey}`);
+    await this.cacheService.setValue(key, JSON.stringify(serializedPCD));
+  }
+
+  /**
+   * Return a response to the feed issuance request. This takes the form of
+   * an array of PCD actions that will be applied to the user's PCD collection.
+   *
+   * @param req The request for feed issuance
+   * @returns The response to the feed issuance request
+   */
+  private async feedIssue(
+    req: PollFeedRequest
+  ): Promise<PollFeedResponseValue> {
+    return traced(LOG_NAME, "feedIssue", async (span) => {
+      if (!req.pcd) {
+        throw new Error("missing credential pcd");
+      }
+
+      if (this.definition.options.paused) {
+        return { actions: [] };
+      }
+
+      const credential =
+        await this.credentialSubservice.verifyAndExpectZupassEmail(req.pcd);
+      const { email, semaphoreId } = credential;
+
+      span?.setAttribute("email", email);
+      span?.setAttribute("semaphore_id", semaphoreId);
+
+      // Consumer is validated, so save them in the consumer list
+      const didUpdate = await this.consumerDB.save(
+        this.id,
+        email,
+        semaphoreId,
+        new Date()
+      );
+      if (didUpdate) {
+        this.consumerDB.save(this.id, email, semaphoreId, new Date());
+      }
+
+      const atomsToIssue = await this.loadMatchingAtoms(credential);
+
+      const serializedPCDs = await Promise.all(
+        atomsToIssue.map(async (atom) => {
+          // Get the final POD entries from the atom, output configuration,
+          // and credential
+          const entries = finalizeAtom(
+            atom,
+            this.definition.options.outputs[atom.outputId],
+            credential
+          );
+
+          const id = uuidv5(serializePODEntries(entries), this.id);
+
+          let serializedPCD = await this.getCachedPCD(id);
+          if (serializedPCD) {
+            return serializedPCD;
+          }
+
+          // Create a PODPCD
+          // @todo handle wrapper PCD outputs
+          const pcd = await PODPCDPackage.prove({
+            entries: {
+              value: entries,
+              argumentType: ArgumentTypeName.Object
+            },
+            privateKey: {
+              value: this.eddsaPrivateKey,
+              argumentType: ArgumentTypeName.String
+            },
+            id: {
+              value: id,
+              argumentType: ArgumentTypeName.String
+            }
+          });
+
+          serializedPCD = await PODPCDPackage.serialize(pcd);
+          await this.storeCachedPCD(id, serializedPCD);
+
+          return serializedPCD;
+        })
+      );
+
+      const actions: PCDAction[] = [];
+
+      actions.push({
+        type: PCDActionType.DeleteFolder,
+        folder: this.definition.options.feedOptions.feedFolder,
+        recursive: true
+      });
```

Comment:
can we add a field within PODPipelineOptions to prevent this DeleteFolder action from being pushed? perhaps just a boolean like `deleteOtherPCDsInFolder`, which can be default true.

this would allow this pipeline to issue PODs to a folder also has PODs not listed in the CSV. for example, this we could reissue old PODs to the 0xPODs folder without deleting the newly "minted" PODs.

Reply:
After the experience with the house PODs, I'd suggest this as a configurable feature for all pipelines.  To be most general, there might be 3 choices: don't clear the folder at all, clear non-recursively, or clear recursively.  Pipelines only really need to clear the folder if they need the ability to "revoke" old PODs/tickets which are removed (not replaced).  If we don't clear the folder we might also start to notice if the PCD IDs of any of our pipelines aren't consistent over time.

Reply:
This can/should be out of scope for this PR, but note that Richard and I apparently both thought that putting a subfolder in the "0xPARC Summer '24" folder was a reasonable thing to do.  Whether we were right or wrong (wrong as it turned out) this suggests to me that at least some users will probably think so too.

Reply:
Also relevant: I think when Arturo re-issues the accidentally-deleted unowned PODs using PODBox, he's going to want to issue them into the same folder as the owned PODs which have been "Minted" using Ahmad's site, so that's a use case for mixing feed and non-feed PCDs in the same folder.  For that case, what he'd want to configure is to skip the DeleteAction entirely (and never grant permission for it) but instead send only Replace actions.

Reply:
For the sake of completeness, this is my best understanding of why some feeds begin by issuing a "delete" command:

We have, in practice, two kinds of feed. One is an append-only feed, like FrogCrypto's feeds, which write PCDs to a folder and never delete anything or replace existing PCDs. This works like an ephemeral stream of PCDs, where the server doesn't maintain any memory of which PCDs have been served to which users. This is why you can't ask the server to restore a frog that you deleted locally - the server simply doesn't have it any more.

The second is the "synchronized folder" type of feed. This is used for tickets, where the aim is to ensure that the client and server agree on a set of facts (viz. what ticket(s) the user has for a given event). Because feed permissioning works in terms of folders, the folder is the unit of synchronization.

The practical reason for deletion of folder contents is that there's no other way to clean up out-dated PCDs that were previously served by the same feed, but which can no longer be identified as such. This could happen where a user has one or more tickets to an event, and one or more of those tickets is deleted by the event organizers. Without clearing the whole folder, there's no way to make sure that the old ticket is removed. The same can also apply in some cases where an update to a ticket causes its ID to change, meaning that the "replace by ID" mechanism can't replace the old ticket with a new one, and instead leaves the users with two ticket PCDs, causing confusing UX when making proofs.

Perhaps turning folder-level deletion off is reasonable in cases where the user is very sure about what they're doing and why. However, "append-only log" and "synchronized folder" are neat mental models that are easy to reason about, and "feed where PCDs are emitted and sometimes they stop being emitted and sometimes they change IDs but you keep all of the old copies forever" doesn't make a lot of sense to me as a product feature.